File: python/src/dexcost/clients.py
"""Wrapper client fallback for environments where monkey-patching won't work (US-021).

Provides ``TrackedOpenAI`` and ``TrackedAnthropic`` -- thin wrappers around
the real SDK clients that auto-record LLM events to the active task context.

Usage::

    from dexcost.clients import TrackedOpenAI

    client = TrackedOpenAI(tracker=tracker)
    # Inside a tracked task, events are auto-recorded:
    response = client.chat.completions.create(model="gpt-4o", messages=[...])
"""

from __future__ import annotations

import logging
from typing import Any

from dexcost.context import get_current_task
from dexcost.models.event import Event
from dexcost.pricing import PricingEngine
from dexcost.storage.protocol import StorageBackend

_log = logging.getLogger(__name__)


def _get_task_id() -> Any:
    """Get the current Task's task_id from context, or None."""
    task = get_current_task()
    if task is None:
        return None
    return task.task_id


class _TrackedOpenAICompletions:
    """Wraps openai chat completions, recording events on each call."""

    def __init__(
        self,
        completions: Any,
        pricing: PricingEngine,
        storage: StorageBackend | None,
    ) -> None:
        self._completions = completions
        self._pricing = pricing
        self._storage = storage

    def create(self, **kwargs: Any) -> Any:
        """Call openai chat.completions.create and record event."""
        response = self._completions.create(**kwargs)
        self._record_event(response, kwargs.get("model", "unknown"))
        return response

    async def acreate(self, **kwargs: Any) -> Any:
        """Call openai async chat.completions.create and record event."""
        response = await self._completions.acreate(**kwargs)
        self._record_event(response, kwargs.get("model", "unknown"))
        return response

    def _record_event(self, response: Any, model: str) -> None:
        task_id = _get_task_id()
        if task_id is None:
            return
        if self._storage is None:
            return

        usage = getattr(response, "usage", None)
        if usage is None:
            return

        input_tokens = getattr(usage, "prompt_tokens", 0) or 0
        output_tokens = getattr(usage, "completion_tokens", 0) or 0
        cached_tokens = getattr(usage, "cached_tokens", 0) or 0

        cost_result = self._pricing.get_cost(model, input_tokens, output_tokens, cached_tokens)

        event = Event(
            task_id=task_id,
            event_type="llm_call",
            cost_usd=cost_result.cost_usd,
            cost_confidence=cost_result.cost_confidence,
            pricing_source=cost_result.pricing_source,
            pricing_version=cost_result.pricing_version,
            provider="openai",
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cached_tokens=cached_tokens,
        )
        self._storage.insert_event(event)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._completions, name)


class _TrackedOpenAIChat:
    """Wraps openai chat namespace."""

    def __init__(
        self,
        chat: Any,
        pricing: PricingEngine,
        storage: StorageBackend | None,
    ) -> None:
        self._chat = chat
        self.completions = _TrackedOpenAICompletions(chat.completions, pricing, storage)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._chat, name)


class TrackedOpenAI:
    """Thin wrapper around ``openai.OpenAI`` that auto-records LLM events.

    Works independently of auto-instrumentation (no monkey-patching).
    Pass an existing client or let the wrapper create one.

    The recommended way is to pass a :class:`~dexcost.tracker.CostTracker`
    via the *tracker* parameter, which provides both pricing and storage::

        from dexcost import CostTracker
        from dexcost.clients import TrackedOpenAI

        tracker = CostTracker(auto_instrument=[])
        client = TrackedOpenAI(tracker=tracker)

    Alternatively, pass *pricing* directly (events are only recorded if
    a tracker with storage is provided).
    """

    def __init__(
        self,
        client: Any = None,
        *,
        tracker: Any = None,
        pricing: PricingEngine | None = None,
        **kwargs: Any,
    ) -> None:
        if client is None:
            import openai

            client = openai.OpenAI(**kwargs)
        self._client = client

        # Resolve pricing and storage from tracker or explicit params
        if tracker is not None:
            self._pricing: PricingEngine = tracker.pricing
            self._storage: StorageBackend | None = tracker.storage
        elif pricing is not None:
            self._pricing = pricing
            self._storage = None
        else:
            self._pricing = PricingEngine(auto_update=False)
            self._storage = None

        self.chat = _TrackedOpenAIChat(client.chat, self._pricing, self._storage)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)


class _TrackedAnthropicMessages:
    """Wraps anthropic messages, recording events on each call."""

    def __init__(
        self,
        messages: Any,
        pricing: PricingEngine,
        storage: StorageBackend | None,
    ) -> None:
        self._messages = messages
        self._pricing = pricing
        self._storage = storage

    def create(self, **kwargs: Any) -> Any:
        """Call anthropic messages.create and record event."""
        response = self._messages.create(**kwargs)
        self._record_event(response, kwargs.get("model", "unknown"))
        return response

    async def acreate(self, **kwargs: Any) -> Any:
        """Call anthropic async messages.create and record event."""
        response = await self._messages.acreate(**kwargs)
        self._record_event(response, kwargs.get("model", "unknown"))
        return response

    def _record_event(self, response: Any, model: str) -> None:
        task_id = _get_task_id()
        if task_id is None:
            return
        if self._storage is None:
            return

        usage = getattr(response, "usage", None)
        if usage is None:
            return

        input_tokens = getattr(usage, "input_tokens", 0) or 0
        output_tokens = getattr(usage, "output_tokens", 0) or 0
        # Anthropic cache tokens
        cache_creation = getattr(usage, "cache_creation_input_tokens", 0) or 0
        cache_read = getattr(usage, "cache_read_input_tokens", 0) or 0
        cached_tokens = cache_creation + cache_read

        cost_result = self._pricing.get_cost(model, input_tokens, output_tokens, cached_tokens)

        event = Event(
            task_id=task_id,
            event_type="llm_call",
            cost_usd=cost_result.cost_usd,
            cost_confidence=cost_result.cost_confidence,
            pricing_source=cost_result.pricing_source,
            pricing_version=cost_result.pricing_version,
            provider="anthropic",
            model=model,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            cached_tokens=cached_tokens,
        )
        self._storage.insert_event(event)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._messages, name)


class TrackedAnthropic:
    """Thin wrapper around ``anthropic.Anthropic`` that auto-records LLM events.

    Works independently of auto-instrumentation (no monkey-patching).

    The recommended way is to pass a :class:`~dexcost.tracker.CostTracker`
    via the *tracker* parameter::

        from dexcost import CostTracker
        from dexcost.clients import TrackedAnthropic

        tracker = CostTracker(auto_instrument=[])
        client = TrackedAnthropic(tracker=tracker)
    """

    def __init__(
        self,
        client: Any = None,
        *,
        tracker: Any = None,
        pricing: PricingEngine | None = None,
        **kwargs: Any,
    ) -> None:
        if client is None:
            import anthropic

            client = anthropic.Anthropic(**kwargs)
        self._client = client

        # Resolve pricing and storage from tracker or explicit params
        if tracker is not None:
            self._pricing: PricingEngine = tracker.pricing
            self._storage: StorageBackend | None = tracker.storage
        elif pricing is not None:
            self._pricing = pricing
            self._storage = None
        else:
            self._pricing = PricingEngine(auto_update=False)
            self._storage = None

        self.messages = _TrackedAnthropicMessages(client.messages, self._pricing, self._storage)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)

File: typescript/src/clients.ts
/**
 * Thin wrapper classes around the OpenAI and Anthropic clients that
 * auto-record LLM cost events.
 *
 * Alternative to auto-instrumentation (no monkey-patching). Matching
 * the Python SDK's clients.py.
 *
 * Usage:
 *
 *   import { TrackedOpenAI } from "@dexcost/sdk/clients";
 *
 *   const client = new TrackedOpenAI({ tracker });
 *   // Inside a tracked task, events are auto-recorded:
 *   const response = await client.chat.completions.create({ model: "gpt-4o", messages: [] });
 */

import { randomUUID } from "node:crypto";
import { getCurrentTask } from "./core/context.js";
import { createCostEvent, Decimal } from "./core/models.js";
import { createAutoTask } from "./core/auto-task.js";
import { registerLlmCapture } from "./core/llm-dedup.js";
import type { Task, CostConfidence, PricingSource } from "./core/models.js";
import { PricingEngine } from "./pricing/engine.js";
import type { EventBuffer } from "./transport/buffer.js";
import type { CostTracker } from "./core/tracker.js";

// ---------------------------------------------------------------------------
// TrackedOpenAI
// ---------------------------------------------------------------------------

/**
 * Wraps `openai.OpenAI` and auto-records llm_call events on every
 * `chat.completions.create` call when running inside a tracked task context.
 */
export class TrackedOpenAI {
  private _client: unknown;
  private _pricing: PricingEngine;
  private _buffer: EventBuffer | null;

  constructor(opts?: { client?: unknown; tracker?: CostTracker; pricing?: PricingEngine }) {
    if (opts?.client !== undefined) {
      this._client = opts.client;
    } else {
      // Lazy require — openai is an optional peer dependency
      let OpenAIClass: new () => unknown;
      try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const openaiModule = require("openai") as { default?: { new(): unknown }; OpenAI?: { new(): unknown } };
        OpenAIClass = (openaiModule.default ?? openaiModule.OpenAI) as typeof OpenAIClass;
      } catch {
        throw new Error(
          "The 'openai' package is required for TrackedOpenAI. Install it with: npm install openai"
        );
      }
      this._client = new OpenAIClass();
    }

    if (opts?.tracker !== undefined) {
      this._pricing = opts.tracker.pricing;
      this._buffer = opts.tracker.buffer;
    } else if (opts?.pricing !== undefined) {
      this._pricing = opts.pricing;
      this._buffer = null;
    } else {
      this._pricing = new PricingEngine();
      this._buffer = null;
    }
  }

  get chat(): { completions: { create: (...args: unknown[]) => Promise<unknown> } } {
    const pricing = this._pricing;
    const buffer = this._buffer;
    const client = this._client as {
      chat: { completions: { create: (...args: unknown[]) => Promise<unknown> } };
    };

    return {
      completions: {
        create: async (...args: unknown[]): Promise<unknown> => {
          const response = await client.chat.completions.create(...args);
          recordOpenAIEvent(response, args[0], pricing, buffer);
          return response;
        },
      },
    };
  }
}

function recordOpenAIEvent(
  response: unknown,
  body: unknown,
  pricing: PricingEngine,
  buffer: EventBuffer | null,
): void {
  if (!buffer) return;

  let task = getCurrentTask();
  if (!task) {
    // Auto-create a task so LLM costs are never silently lost
    // (mirrors Python create_auto_task).
    task = createAutoTask("openai.chat");
    buffer.upsertTask(task);
  }

  const resp = response as Record<string, unknown>;
  const bodyObj = (body ?? {}) as Record<string, unknown>;

  const model: string =
    (resp["model"] as string | undefined) ??
    (bodyObj["model"] as string | undefined) ??
    "unknown";

  const usage = resp["usage"] as Record<string, unknown> | undefined;
  const hasUsage = usage != null;

  const inputTokens: number = (usage?.["prompt_tokens"] as number | undefined) ?? 0;
  const outputTokens: number = (usage?.["completion_tokens"] as number | undefined) ?? 0;
  const promptDetails = usage?.["prompt_tokens_details"] as Record<string, unknown> | undefined;
  const cachedTokens: number = (promptDetails?.["cached_tokens"] as number | undefined) ?? 0;

  let costUsd: Decimal = new Decimal(0);
  let costConfidence: CostConfidence = "estimated";
  let pricingSource: PricingSource = "unknown";

  if (hasUsage) {
    const result = pricing.getCost(model, inputTokens, outputTokens, cachedTokens);
    costUsd = result.costUsd;
    costConfidence = result.costConfidence;
    pricingSource = result.pricingSource;
  }

  _addEventAndUpdateTask(task, buffer, {
    provider: "openai",
    model,
    inputTokens,
    outputTokens,
    cachedTokens,
    costUsd,
    costConfidence,
    pricingSource,
  });
}

// ---------------------------------------------------------------------------
// TrackedAnthropic
// ---------------------------------------------------------------------------

/**
 * Wraps `@anthropic-ai/sdk Anthropic` and auto-records llm_call events on
 * every `messages.create` call when running inside a tracked task context.
 */
export class TrackedAnthropic {
  private _client: unknown;
  private _pricing: PricingEngine;
  private _buffer: EventBuffer | null;

  constructor(opts?: { client?: unknown; tracker?: CostTracker; pricing?: PricingEngine }) {
    if (opts?.client !== undefined) {
      this._client = opts.client;
    } else {
      // Lazy require — @anthropic-ai/sdk is an optional peer dependency
      let AnthropicClass: new () => unknown;
      try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const anthropicModule = require("@anthropic-ai/sdk") as {
          default?: { new(): unknown };
          Anthropic?: { new(): unknown };
        };
        AnthropicClass = (anthropicModule.default ?? anthropicModule.Anthropic) as typeof AnthropicClass;
      } catch {
        throw new Error(
          "The '@anthropic-ai/sdk' package is required for TrackedAnthropic. Install it with: npm install @anthropic-ai/sdk"
        );
      }
      this._client = new AnthropicClass();
    }

    if (opts?.tracker !== undefined) {
      this._pricing = opts.tracker.pricing;
      this._buffer = opts.tracker.buffer;
    } else if (opts?.pricing !== undefined) {
      this._pricing = opts.pricing;
      this._buffer = null;
    } else {
      this._pricing = new PricingEngine();
      this._buffer = null;
    }
  }

  get messages(): { create: (...args: unknown[]) => Promise<unknown> } {
    const pricing = this._pricing;
    const buffer = this._buffer;
    const client = this._client as {
      messages: { create: (...args: unknown[]) => Promise<unknown> };
    };

    return {
      create: async (...args: unknown[]): Promise<unknown> => {
        const response = await client.messages.create(...args);
        recordAnthropicEvent(response, args[0], pricing, buffer);
        return response;
      },
    };
  }
}

function recordAnthropicEvent(
  response: unknown,
  body: unknown,
  pricing: PricingEngine,
  buffer: EventBuffer | null,
): void {
  if (!buffer) return;

  let task = getCurrentTask();
  if (!task) {
    // Auto-create a task so LLM costs are never silently lost
    // (mirrors Python create_auto_task).
    task = createAutoTask("anthropic.messages");
    buffer.upsertTask(task);
  }

  const resp = response as Record<string, unknown>;
  const bodyObj = (body ?? {}) as Record<string, unknown>;

  const model: string =
    (resp["model"] as string | undefined) ??
    (bodyObj["model"] as string | undefined) ??
    "unknown";

  const usage = resp["usage"] as Record<string, unknown> | undefined;
  const hasUsage = usage != null;

  const inputTokens: number = (usage?.["input_tokens"] as number | undefined) ?? 0;
  const outputTokens: number = (usage?.["output_tokens"] as number | undefined) ?? 0;
  const cacheCreationTokens: number =
    (usage?.["cache_creation_input_tokens"] as number | undefined) ?? 0;
  const cacheReadTokens: number =
    (usage?.["cache_read_input_tokens"] as number | undefined) ?? 0;
  // cachedTokens tracks cache_read (for aggregation); cache_creation stored in details
  const cachedTokens: number = cacheReadTokens;

  let costUsd: Decimal = new Decimal(0);
  let costConfidence: CostConfidence = "estimated";
  let pricingSource: PricingSource = "unknown";

  if (hasUsage) {
    const result = pricing.getCost(
      model,
      inputTokens,
      outputTokens,
      cacheReadTokens,
      cacheCreationTokens,
    );
    costUsd = result.costUsd;
    costConfidence = result.costConfidence;
    pricingSource = result.pricingSource;
  }

  _addEventAndUpdateTask(task, buffer, {
    provider: "anthropic",
    model,
    inputTokens,
    outputTokens,
    cachedTokens,
    costUsd,
    costConfidence,
    pricingSource,
    details: cacheCreationTokens > 0 ? { cache_creation_input_tokens: cacheCreationTokens } : undefined,
  });
}

// ---------------------------------------------------------------------------
// Shared helper
// ---------------------------------------------------------------------------

interface EventSpec {
  provider: string;
  model: string;
  inputTokens: number;
  outputTokens: number;
  cachedTokens: number;
  costUsd: Decimal;
  costConfidence: CostConfidence;
  pricingSource: PricingSource;
  details?: Record<string, unknown>;
}

function _addEventAndUpdateTask(
  task: Task,
  buffer: EventBuffer,
  spec: EventSpec,
): void {
  const event = createCostEvent({
    eventId: randomUUID(),
    taskId: task.taskId,
    eventType: "llm_call",
    costUsd: spec.costUsd,
    costConfidence: spec.costConfidence,
    pricingSource: spec.pricingSource,
    provider: spec.provider,
    model: spec.model,
    inputTokens: spec.inputTokens,
    outputTokens: spec.outputTokens,
    cachedTokens: spec.cachedTokens,
    isRetry: false,
    details: spec.details ?? {},
  });

  buffer.addEvent(event);
  registerLlmCapture(task.taskId, event.inputTokens ?? 0, event.outputTokens ?? 0);

  task.llmCostUsd = task.llmCostUsd.plus(spec.costUsd);
  task.totalCostUsd = task.totalCostUsd.plus(spec.costUsd);
  task.totalInputTokens += spec.inputTokens;
  task.totalOutputTokens += spec.outputTokens;
  task.totalCachedTokens += spec.cachedTokens;

  buffer.upsertTask(task);
}

// ---------------------------------------------------------------------------
// wrap* convention entry points
// ---------------------------------------------------------------------------

/**
 * Wrap an OpenAI client instance for cost tracking (ecosystem `wrapOpenAI`
 * convention). Returns a {@link TrackedOpenAI} exposing the chat-completions
 * surface; for FULL client-surface coverage prefer injecting a tracked
 * fetch instead: `new OpenAI({ fetch: createDexcostFetch() })`.
 */
export function wrapOpenAI(
  client: unknown,
  opts?: { tracker?: CostTracker },
): TrackedOpenAI {
  return new TrackedOpenAI({ client, tracker: opts?.tracker });
}

/**
 * Wrap an Anthropic client instance for cost tracking (ecosystem
 * `wrapAnthropic` convention). Returns a {@link TrackedAnthropic} exposing
 * the messages surface; for FULL client-surface coverage prefer injecting
 * a tracked fetch instead: `new Anthropic({ fetch: createDexcostFetch() })`.
 */
export function wrapAnthropic(
  client: unknown,
  opts?: { tracker?: CostTracker },
): TrackedAnthropic {
  return new TrackedAnthropic({ client, tracker: opts?.tracker });
}
